jcapi-commands.go

```diff
@@ -0,0 +1,158 @@
+	ScheduledRunDate string   `json:"scheduledRunDate"`         // when LaunchType='scheduled', set to the date on which to start the command
+	ScheduledRunTime string   `json:"scheduledRunTime"`         // when LaunchType='scheduled', set to the time at which to start the command
+	Trigger          string   `json:"trigger,omitempty"`        // generate trigger (No longer supported)
+	Timeout          string   `json:"timeout"`                  // Command time out in seconds, after which it will be killed
+	Organization     string   `json:"organization,omitempty"`   // organization ID for this command (auto-populated)
+	Sudo             bool     `json:"sudo"`                     // Indicates whether the command should be run with sudo
+
+	Skip  int `json:"skip"`  // Objects to skip on /search POST
+	Limit int `json:"limit"` // Max objects to return on /search POST
+}
+
+func (e JCCommand) ToString() string {
+	return fmt.Sprintf("command: %v", e)
+}
+
+func getJCCommandsFromResults(result []byte) (commands []JCCommand, err JCError) {
+	commandResults := JCCommandResults{}
+
+	err = json.Unmarshal(result, &commandResults)
+	if err != nil {
+		err = fmt.Errorf("Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+	}
+
+	commands = commandResults.Results
+
+	return
+}
+
+func (jc JCAPI) GetAllCommands() (commandList []JCCommand, err JCError) {
+	var empty []byte
+
+	for skip := 0; skip == 0 || len(commandList) == searchLimit; skip += searchSkipInterval {
+		url := fmt.Sprintf("%s?sort=hostname&skip=%d&limit=%d", COMMAND_PATH, skip, searchLimit)
+
+		jcSysRec, err2 := jc.DoBytes(MapJCOpToHTTP(Read), url, empty)
+
+		if err2 != nil {
+			return nil, fmt.Errorf("ERROR: Get commands to JumpCloud failed, err='%s'", err2)
+		}
+
+		if jcSysRec == nil {
+			return nil, fmt.Errorf("ERROR: No commands found")
+		}
+
+		resultsBlock, err2 := getJCCommandsFromResults(jcSysRec)
+		if err2 != nil {
+			err = fmt.Errorf("Could not get resultsBlock data, err='%s'", err2.Error())
+			return
+		}
+
+		for i, _ := range resultsBlock {
+			if resultsBlock[i].Id != "" {
+				commandList = append(commandList, resultsBlock[i])
+			}
+		}
+
+	}
+
+	return
+}
+
+func FindCommandById(commands []JCCommand, id string) (result *JCCommand, index int) {
+	index = FindObject(GetInterfaceArrayFromJCCommand(commands), "Id", id)
+	if index >= 0 {
+		result = &commands[index]
+	}
+
+	return
+}
+
+func GetInterfaceArrayFromJCCommand(commands []JCCommand) (interfaceArray []interface{}) {
+	interfaceArray = make([]interface{}, len(commands), len(commands))
+
+	for i := range commands {
+		interfaceArray[i] = commands[i]
+	}
+
+	return
+}
+
+//
+// Add or Update a command in place on JumpCloud
+//
+func (jc JCAPI) AddUpdateCommand(op JCOp, command JCCommand) (id string, err JCError) {
+	id, err = jc.HandleCommand(COMMAND_PATH, op, command)
+
+	return
+}
```

Comment:
To keep the API consistent, I think I'd split this into `AddCommand` and `UpdateCommand` and have them pass the correct JCOp into `HandleCommand`. Otherwise I'd argue that you should check here that `op` is `Insert` or `Update`, and if not return an error.  But having two ways to call command crud in here doesn't seem worth while.